Garbage collection

yhr opened the pull request:
Add an option to enable garbage collection in zenfs.

The garbage collection, if enabled, kicks in under 20% capacity left and cleans out zones with 1% valid data in them to start off with, increasing to 40% at 99% capacity.

This scheme reduces space amplification down to as little as 5% (from ~10-15%) in my testing.

MaisenbacherD commented on fs/fs_zenfs.cc:
```diff
@@ -255,11 +256,65 @@ ZenFS::~ZenFS() {
+    for (const auto& zone : snapshot.zones_) {
+      if (zone.capacity == 0) {
+        uint64_t garbage_percent =
+            100 - 100 * zone.used_capacity / zone.max_capacity;
```
Maybe we should consider refactoring `zone.used_capacity` to something like `zone.valid_data_capacity`.

yhr replied:
Let's think about naming and do that as a separate patch

MaisenbacherD commented on fs/fs_zenfs.cc:
```diff
@@ -255,11 +256,65 @@ ZenFS::~ZenFS() {
+
+    usleep(1000 * 1000 * 10);
+
+    if (free_percent > GC_START_LEVEL) continue;
```
Maybe we should add another threshold that prevents reclaiming zones if there is not much reclaimable space.

yhr replied:
The code is already doing that, and it depends on how much space is available on the disk

metaspace commented on fs/fs_zenfs.cc:
```diff
@@ -255,11 +256,65 @@ ZenFS::~ZenFS() {
+      if (zone.capacity == 0) {
+        uint64_t garbage_percent =
+            100 - 100 * zone.used_capacity / zone.max_capacity;
+        if (garbage_percent > threshold && garbage_percent < 100) {
```
@MaisenbacherD  If `garbage_percent == 100`, no migration is required. The zone will be picked up by `ZonedBlockDevice::ResetUnusedIOZones()` at some point.

metaspace commented on fs/fs_zenfs.cc:
```diff
@@ -253,11 +256,65 @@ ZenFS::~ZenFS() {
+    std::set<uint64_t> migrate_zone_ids;
+    for (const auto& zone : snapshot.zones_) {
+      if (zone.capacity == 0) {
+        uint64_t garbage_percent =
```
```suggestion
        uint64_t garbage_percent_approximate =
```

The snapshots are not consistent in the face of concurrency. It sort of bothers me, but I don't think it will break anything. Just be aware that the numbers might not add up.

metaspace reviewed (changes requested):
I don't think it is safe to call `MigrateExtents` concurrently like this. We would need to make `MigrateExtents` thread safe or stop the world while migrating.

yhr commented on fs/fs_zenfs.cc:
```diff
@@ -255,11 +256,65 @@ ZenFS::~ZenFS() {
 
+  if (gc_worker_ != nullptr) {
+    run_gc_worker_ = false;
+    gc_worker_->join();
```
Yep

yhr commented on fs/fs_zenfs.cc:
```diff
@@ -253,11 +256,65 @@ ZenFS::~ZenFS() {
+      IOStatus s;
+      Info(logger_, "Garbage collecting %d extents \n",
+           (int)migrate_exts.size());
+      s = MigrateExtents(migrate_exts);
```
MigrateExtents should be thread safe, as long as the file is not being written to in parallel (which we can't guarantee at the moment, so we need to ensure that re-opens and migration is mutually exclusive)

yhr replied:
There is actually one thing that looks fishy in MigrateExtents, the ReplaceExtents locking looks broken, we need to make sure that readers won't read outdated information

yhr replied:
I checked and ReplaceExtents looks ok, read/write locking was introduced as part of #123

yhr commented:
I've addressed all comments(except the refactoring things I believe) and increased the gc aggressiveness a bit (slope 2 -> 3). It looks stable and performing nicely, (sysbench and db_bench tests give promising results), but it might need more fine tuning, so let's keep this on a branch until we've finished evaluating this,

Haltz commented:
Hi guys! A few days ago I tested this garbage collection draft and it stuck. I guess the GC threshold is not good. Considering a case where all zones have valid and invalid data at the same time, ZenFS maybe can not find a empty zone to collect the garbage. Will you take a look at this?
Please tell me if my assumption is wrong! Thanks very much!

aravind-wdc commented:
> Hi guys! A few days ago I tested this garbage collection draft and it stuck. I guess the GC threshold is not good. Considering a case where all zones have valid and invalid data at the same time, ZenFS maybe can not find a empty zone to collect the garbage. Will you take a look at this? Please tell me if my assumption is wrong! Thanks very much!

Hi Haltz,

Thanks for testing it out.
Could you give more information on what happened with the hang(how to reproduce this? ) ? 
What were you testing on it with (Hardware) ?
What application were you running ? Did you look at the logs at /tmp/znefs_*.log ?

Haltz commented:
I will give more infomation(maybe a few days later). I was testing this branch on FEMU(a emulator based on QEMU). RocksDB was in the guest virtual machine and using a emulated ZNS SSD as backend storage device.

MaisenbacherD commented on fs/fs_zenfs.h:
```diff
@@ -452,6 +456,12 @@ class ZenFS : public FileSystemWrapper {
       const std::vector<ZoneExtentSnapshot*>& migrate_exts);
+
+ private:
+  const uint64_t GC_START_LEVEL =
+      20;                      /* Enable GC when < 20% free space available */
+  const uint64_t GC_SLOPE = 3; /* GC agressiveness */
```
It would be nice to be able to overwrite the default `GC_SLOPE` and `GC_START_LEVEL` though the ZenFS utility in mkfs.

MaisenbacherD reviewed (commented):
The GC feature itself looks good to me :)
However, I thing we need to add a commit that fixes the `MigrateFileExtents` as mentioned in the inline comment.

metaspace commented on fs/fs_zenfs.cc:
```diff
@@ -256,11 +259,67 @@ ZenFS::~ZenFS() {
+
+    uint64_t non_free = zbd_->GetUsedSpace() + zbd_->GetReclaimableSpace();
+    uint64_t free = zbd_->GetFreeSpace();
+    uint64_t free_percent = (100 * free) / (free + non_free);
```
Maybe it would make sense to use floats here? The integer arithmetic is going to be a little bit off.

MaisenbacherD commented on fs/fs_zenfs.cc:
```diff
@@ -258,11 +259,67 @@ ZenFS::~ZenFS() {
+      IOStatus s;
+      Info(logger_, "Garbage collecting %d extents \n",
+           (int)migrate_exts.size());
+      s = MigrateExtents(migrate_exts);
```
As we discussed, `ResetUnusedIOZones` does not clean zones which where not migrated by not acquiring the WR lock beforehand.

yhr commented:
I have some follow-up patches to this pull request(including migration bug fixes), but will merge this and then submit the rest as separate PRs.
